### PR TITLE
Point Set Processing: Use Named Parameters 

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

[Pre-approved small feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Named_Parameters_for_PSP) that changes the API of PSP from iterators/optional parameters with overloads to range/named parameters.

## Release Management

* Affected package(s): Point Set Processing, Point Set 3
* Small Feature: [Features/Small_Features/Named_Parameters_for_PSP](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Named_Parameters_for_PSP)
* pre-approved --Andreas Fabri 15:04, 20 December 2017 (CET)

### CODE REVIEW

**Comment by sloriot** [comment]
Which release do you target?

**Comment by sgiraudot** [comment]
4.12 if possible…

**Comment by sloriot** [comment]
@sgiraudot Travis is failing

File: BGL/include/CGAL/boost/graph/named_params_helper.h [hunk 1]
```diff
@@ -102,111 +110,255 @@ namespace CGAL {
...
-    DefaultVPMap;
-  public:
-    typedef typename boost::lookup_named_param_def<
-    internal_np::vertex_point_t,
-    NamedParameters,
-    DefaultVPMap
-    > ::type  type;
-    typedef typename boost::lookup_named_param_def<
+  
+    template<typename PolygonMesh, typename NamedParameters>
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/named_params_helper.h]: no need to change the indentation when you add a namespace (in addition it creates conflicts when merging other branches)

**Comment by sloriot** [comment]
I have errors when I try to compile it locally:

```
/home/sloriot/CGAL/git/integration/Point_set_3/examples/Point_set_3/point_set_algo.cpp:43:9: error: expected primary-expression before ‘;’ token
      12,; // Number of neighbors
/home/sloriot/CGAL/git/integration/Point_set_3/examples/Point_set_3/point_set_algo.cpp:45:23: error: expected ‘;’ before ‘)’ token
      degree_fitting(2));     // additional named parameter specific to jet_estimate_normals
/home/sloriot/CGAL/git/integration/Point_set_processing_3/examples/Point_set_processing_3
```

**Comment by sloriot** [comment]
After your last commit, I'm now have:

```
/home/sloriot/CGAL/git/integration/Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction.cpp:199:39: error: ‘make_normal_of_point_with_normal_pmap’ is not a member of ‘CGAL’
/home/sloriot/CGAL/git/integration/Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction.cpp:255:37: error: ‘make_normal_of_point_with_normal_pmap’ is not a member of ‘CGAL’
/home/sloriot/CGAL/git/integration/Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp:50:37: error: ‘make_normal_of_point_with_normal_pmap’ is not a member of ‘CGAL’
/home/sloriot/CGAL/git/integration/Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp:62:52: error: ‘make_normal_of_point_with_normal_pmap’ is not a member of ‘CGAL’
```

**Comment by lrineau** [comment]
Lookup issue detected by Travis: https://travis-ci.org/CGAL/cgal/jobs/320229143

**Comment by lrineau** [comment]
This PR is probably the cause of several yellow and red lines in the testsuite:
- [Advancing_front_surface_reconstruction_Examples](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Advancing_front_surface_reconstruction_Examples)
- [Classification](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Classification) and Classification_Examples
- [Point_set_shape_detection_3_Examples](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Point_set_shape_detection_3_Examples)
- [Triangulation_3_Demo](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Triangulation_3_Demo)

**Comment by sloriot** [comment]
And  [Polygon_mesh_processing](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Polygon_mesh_processing) IMO.

**Comment by lrineau** [comment]
> And  [Polygon_mesh_processing](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-133.shtml#Polygon_mesh_processing) IMO.
 
Probably.

By the way, the name `GetVertexPointMap` of the functor does not follow our naming rules: it should be renamed `Get_vertex_point_map`.

**Comment by lrineau** [comment]
Before 3676e29, `GetVertexPointMap` was in the `CGAL` namespace, and it was moved into the `CGAL::Polygon_mesh_processing` namespace. That is the cause of the compilation error.

And 3676e29 is part of this PR.

**Comment by sgiraudot** [comment]
Okay, I fixed the errors/warnings. Few comments:
* Warnings in classification are actually due to PR https://github.com/CGAL/cgal/pull/2690 (I fixed them on the branch)
* Warnings in Shape Detection come from the combination of this PR with PR https://github.com/CGAL/cgal/pull/2695, so depending on which one will be merged first, I'll have to fix the problem afterwards…

(Also, I had to rebase to get the errors in PMP.)

**Comment by lrineau** [comment]
Optimal_transportation_reconstruction_2_Demo has warnings about using deprecated functions, in [CGAL-4.12-Ic-136](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-136.shtml#Optimal_transportation_reconstruction_2_Demo). Maybe that is due to this PR.

**Comment by lrineau** [comment]
https://github.com/CGAL/cgal/pull/2694#issuecomment-355926879
> Okay, I fixed the errors/warnings. Few comments:
> * Warnings in classification are actually due to PR https://github.com/CGAL/cgal/pull/2690 (I fixed them on the branch)
> * Warnings in Shape Detection come from the combination of this PR with PR ?
https://github.com/CGAL/cgal/pull/2695, so depending on which one will be merged first, I'll have to fix the problem afterwards…


I suggest you assume both PR will be merged into the same batch of merges. You can fix the warnings in one of the two PRs. Otherwise we cannot rely on the testsuite results.

Or you choose one PR that has to be merged before the other one, and I will only test that one. Let me know.

**Comment by sgiraudot** [comment]
I rebase both the Shape Detection and Optimal Transportation branches on this one.

**Comment by lrineau** [comment]
This PR also has the conflict on `Point_set_processing_3/doc/Point_set_processing_3/Point_set_processing_3.txt`!

**Comment by sgiraudot** [comment]
(`apt-get` failure for Travis.)

**Comment by lrineau** [comment]
And now it is green.

**Comment by sgiraudot** [comment]
Still some errors in the testsuite, I'll look into them.

**Pull request merged**